### PR TITLE
Oracle - Implementation of Batch Insertion

### PR DESCRIPTION
**Pull request by ralmsdeveloper**
- [x] Commit messages follow this format

Implementation of the batch insert, no test unit was created for this implementation.

At Oracle, we have some limitation when it comes to the return of the ResultSet.

But this implementation is functional and we can improve.

Forgive my English, I'm Brazilian and I do not speak well.

But this is my humble contribution.

**With this, we will have something like this:**

```sql
DECLARE
TYPE efRowProduct IS RECORD
(
    Id NUMBER(10)
);
TYPE efProduct IS TABLE OF efRowProduct;
listProduct efProduct;
BEGIN

INSERT INTO "Person" ("Name", "LastName")
VALUES (:p0, :p1);

listProduct := efProduct();
listProduct.extend(2);

INSERT INTO "Product" ("Description", "Price", "Quanty")
VALUES (:p2, :p3, :p4)
RETURNING "Id" INTO listProduct(1);

INSERT INTO "Product" ("Description", "Price", "Quanty")
VALUES (:p5, :p6, :p7)
RETURNING "Id" INTO listProduct(2);

OPEN :cur1 FOR SELECT listProduct(1).Id FROM DUAL;
OPEN :cur2 FOR SELECT listProduct(2).Id FROM DUAL;
END;
````

- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Code meets the expectations our engineering guidelines. https://github.com/aspnet/Home/wiki/Engineering-guidelines.

### CODE REVIEW

**Comment by ralmsdeveloper** [comment]
@ajcvickers  , can you see if it's useful?

**Comment by ralmsdeveloper** [comment]
Okay, I'll be doing this.

File: samples/OracleProvider/src/OracleProvider/Storage/Internal/OracleRelationalCommandBuilderFactory.cs [hunk 1]
```diff
@@ -315,14 +316,18 @@ private static string AdjustCommandText(string commandText)
                     }
 
                     // HACK: Need to make it easier to add this in update pipeline.
-                    if (command.CommandText.Contains(":cur"))
+                    if (command.CommandText.IndexOf(":cur") > 0)
                     {
-                        command.Parameters.Add(
+                        var cursors = Regex.Matches(command.CommandText, ":cur").Count;
```
Comment (smitpatel) [inline samples/OracleProvider/src/OracleProvider/Storage/Internal/OracleRelationalCommandBuilderFactory.cs]: This would need timeout on regex prolly.

Comment (ralmsdeveloper) [reply samples/OracleProvider/src/OracleProvider/Storage/Internal/OracleRelationalCommandBuilderFactory.cs]: True, I'm implementing!
Tanks @smitpatel

File: samples/OracleProvider/src/OracleProvider/Update/Internal/IOracleUpdateSqlGenerator.cs [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+using System.Collections.Generic; 
+using System.Text; 
+
+namespace Microsoft.EntityFrameworkCore.Update.Internal
+{
+	public interface IOracleUpdateSqlGenerator : IUpdateSqlGenerator
+	{
+		ResultSetMapping AppendBulkInsertOperation(
+			StringBuilder commandStringBuilder,
```
Comment (anpete) [inline samples/OracleProvider/src/OracleProvider/Update/Internal/IOracleUpdateSqlGenerator.cs]: Use 4 spaces for indenting

File: samples/OracleProvider/src/OracleProvider/Update/Internal/OracleModificationCommandBatch.cs [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.EntityFrameworkCore.Update.Internal
+{
+    public class OracleModificationCommandBatch : AffectedCountModificationCommandBatch
+    {
+		internal const int MaxParameterCount = 1000;
+		internal const int MaxRowCount = 300;
+		internal int _CountParameter = 1;
```
Comment (anpete) [inline samples/OracleProvider/src/OracleProvider/Update/Internal/OracleModificationCommandBatch.cs]: Why are these internal?

File: samples/OracleProvider/src/OracleProvider/Update/Internal/OracleModificationCommandBatch.cs [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+			_BatchInsertCommands = new List<ModificationCommand>();
+			_VariablesCommand = new StringBuilder();
+		}
+
+		protected new virtual IOracleUpdateSqlGenerator UpdateSqlGenerator => (IOracleUpdateSqlGenerator)base.UpdateSqlGenerator;
+
+		protected override bool CanAddCommand(ModificationCommand modificationCommand)
+		{
+			if (ModificationCommands.Count >= MaxRowCount)
+				return false;
```
Comment (anpete) [inline samples/OracleProvider/src/OracleProvider/Update/Internal/OracleModificationCommandBatch.cs]: Force braces even for single statement blocks

File: samples/OracleProvider/src/OracleProvider/Storage/Internal/OracleRelationalCommandBuilderFactory.cs (hunk 1 above)
Comment (anpete) [reply samples/OracleProvider/src/OracleProvider/Storage/Internal/OracleRelationalCommandBuilderFactory.cs]: Perhaps @AndriySvyryd can help us to remove this hack.

File: samples/OracleProvider/src/OracleProvider/Update/Internal/OracleModificationCommandBatch.cs (hunk 3 above)
Comment (anpete) [reply samples/OracleProvider/src/OracleProvider/Update/Internal/OracleModificationCommandBatch.cs]: Fields are _lowerCamelCase

File: samples/OracleProvider/test/OracleProvider.FunctionalTests/OracleInsertBatch.cs [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+using System.Linq;
+using System.Text;
+using Xunit;
+
+namespace Microsoft.EntityFrameworkCore
+{
+    public class OracleInsertBatch
+    {
+        [Fact]
+        public void insert_custormer_product_batch()
```
Comment (anpete) [inline samples/OracleProvider/test/OracleProvider.FunctionalTests/OracleInsertBatch.cs]: Typo and capitalize

File: samples/OracleProvider/src/OracleProvider/Update/Internal/OracleModificationCommandBatch.cs (hunk 3 above)
Comment (ralmsdeveloper) [reply samples/OracleProvider/src/OracleProvider/Update/Internal/OracleModificationCommandBatch.cs]: Done!

File: samples/OracleProvider/src/OracleProvider/Update/Internal/OracleModificationCommandBatch.cs (hunk 4 above)
Comment (ralmsdeveloper) [reply samples/OracleProvider/src/OracleProvider/Update/Internal/OracleModificationCommandBatch.cs]: Done!

File: samples/OracleProvider/src/OracleProvider/Update/Internal/IOracleUpdateSqlGenerator.cs (hunk 2 above)
Comment (ralmsdeveloper) [reply samples/OracleProvider/src/OracleProvider/Update/Internal/IOracleUpdateSqlGenerator.cs]: Done!

File: samples/OracleProvider/test/OracleProvider.FunctionalTests/OracleInsertBatch.cs (hunk 5 above)
Comment (ralmsdeveloper) [reply samples/OracleProvider/test/OracleProvider.FunctionalTests/OracleInsertBatch.cs]: Thanks @anpete.
Done!

File: samples/OracleProvider/src/OracleProvider/Storage/Internal/OracleRelationalCommandBuilderFactory.cs (hunk 1 above)
Comment (ralmsdeveloper) [reply samples/OracleProvider/src/OracleProvider/Storage/Internal/OracleRelationalCommandBuilderFactory.cs]: @anpete , I understand, but when the PR started was done, I was curious about this, but then I understood the initial idea for this.
I think at the moment, wearing that is good.
We can improve later, I can be responsible for this improvement.

I'll see what features can improve with that. 
But at the moment I think this is very functional.

File: samples/OracleProvider/src/OracleProvider/Storage/Internal/OracleRelationalCommandBuilderFactory.cs [hunk 6]
```diff
@@ -315,14 +316,19 @@ private static string AdjustCommandText(string commandText)
                     }
 
                     // HACK: Need to make it easier to add this in update pipeline.
-                    if (command.CommandText.Contains(":cur"))
+                    if (command.CommandText.IndexOf(":cur") > 0)
                     {
-                        command.Parameters.Add(
+                        var cursors = Regex.Matches(command.CommandText, ":cur",
+                                                    RegexOptions.Compiled, TimeSpan.FromSeconds(2)).Count;
```
Comment (smitpatel) [inline samples/OracleProvider/src/OracleProvider/Storage/Internal/OracleRelationalCommandBuilderFactory.cs]: Please use `TimeSpan.FromMilliseconds(1000.0)` that's what we are using in whole codebase for timeout.

Comment (ralmsdeveloper) [reply samples/OracleProvider/src/OracleProvider/Storage/Internal/OracleRelationalCommandBuilderFactory.cs]: Okay, I'm doing it!

File: samples/OracleProvider/src/OracleProvider/Update/Internal/OracleModificationCommandBatch.cs [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+
+		protected override int GetParameterCount() => _countParameter;
+
+		private static int CountParameters(ModificationCommand modificationCommand)
+		{
+			var parameterCount = 0;
+			foreach (var columnModification in modificationCommand.ColumnModifications)
+			{
+				if (columnModification.UseCurrentValueParameter)
+					parameterCount++;
```
Comment (AndriySvyryd) [inline samples/OracleProvider/src/OracleProvider/Update/Internal/OracleModificationCommandBatch.cs]: Use braces for single-line blocks

File: samples/OracleProvider/src/OracleProvider/Update/Internal/OracleModificationCommandBatch.cs [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+
+				if (columnModification.UseOriginalValueParameter)
+					parameterCount++;
+			}
+			return parameterCount;
```
Comment (AndriySvyryd) [inline samples/OracleProvider/src/OracleProvider/Update/Internal/OracleModificationCommandBatch.cs]: Shouldn't the `cur` parameters be counted as well?

File: samples/OracleProvider/src/OracleProvider/Storage/Internal/OracleRelationalCommandBuilderFactory.cs [hunk 9]
```diff
@@ -315,14 +316,19 @@ private static string AdjustCommandText(string commandText)
                     }
 
                     // HACK: Need to make it easier to add this in update pipeline.
-                    if (command.CommandText.Contains(":cur"))
+                    if (command.CommandText.IndexOf(":cur") > 0)
```
Comment (AndriySvyryd) [inline samples/OracleProvider/src/OracleProvider/Storage/Internal/OracleRelationalCommandBuilderFactory.cs]: These parameters should be added in `OracleModificationCommandBatch`. You could split `ReaderModificationCommandBatch.CreateStoreCommand()` to make it easier to add parameters in this way.

File: samples/OracleProvider/test/OracleProvider.FunctionalTests/OracleInsertBatch.cs [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+             .AddEntityFrameworkOracle()
+             .BuildServiceProvider();
+              
+            using (var store = OracleTestStore.GetNorthwindStore())
+            {
+                using (var connection = new OracleConnection(store.ConnectionString))
+                {
+                    BatchInsertContext cxt = new BatchInsertContext(serviceProvider, connection);
+                    //Test Insert Batch Customer
+                    for (int i = 0; i < 20; i++)
```
Comment (AndriySvyryd) [inline samples/OracleProvider/test/OracleProvider.FunctionalTests/OracleInsertBatch.cs]: You should insert enough entities to trigger the max parameter count check

File: samples/OracleProvider/test/OracleProvider.FunctionalTests/OracleInsertBatch.cs [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+                            CompanyName = $"Net Foundation Test {i}",
+                            Fax = "79 XXXX-8693"
+                        });
+                    }
+
+                    Assert.Equal(20, cxt.SaveChanges());
+
+                    //Test Insert Batch Product
+                    CreateTableProduct(cxt);
+                    for (int i = 0; i < 50; i++)
```
Comment (AndriySvyryd) [inline samples/OracleProvider/test/OracleProvider.FunctionalTests/OracleInsertBatch.cs]: You should insert enough entities to trigger the max parameter count check

File: samples/OracleProvider/src/OracleProvider/Storage/Internal/OracleRelationalCommandBuilderFactory.cs (hunk 6 above)
Comment (ralmsdeveloper) [reply samples/OracleProvider/src/OracleProvider/Storage/Internal/OracleRelationalCommandBuilderFactory.cs]: Done!

File: samples/OracleProvider/src/OracleProvider/Storage/Internal/OracleRelationalCommandBuilderFactory.cs (hunk 9 above)
Comment (ralmsdeveloper) [reply samples/OracleProvider/src/OracleProvider/Storage/Internal/OracleRelationalCommandBuilderFactory.cs]: I made some adjustments, there was already this class, but I see it!

File: samples/OracleProvider/test/OracleProvider.FunctionalTests/OracleInsertBatch.cs (hunk 11 above)
Comment (ralmsdeveloper) [reply samples/OracleProvider/test/OracleProvider.FunctionalTests/OracleInsertBatch.cs]: Ok I'm doing this, at the moment the intention was to just run a simple test!

File: samples/OracleProvider/src/OracleProvider/Storage/Internal/OracleRelationalCommandBuilderFactory.cs (hunk 9 above)
Comment (ralmsdeveloper) [reply samples/OracleProvider/src/OracleProvider/Storage/Internal/OracleRelationalCommandBuilderFactory.cs]: Done!

File: samples/OracleProvider/src/OracleProvider/Update/Internal/OracleModificationCommandBatch.cs [hunk 12]
```diff
@@ -0,0 +1,351 @@
...
+                    nameParameter,
+                    new OracleParameter(
+                      nameParameter,
+                       OracleDbType.RefCursor,
+                       DBNull.Value,
+                       ParameterDirection.Output));
+            }
+
+
+
```
Comment (AndriySvyryd) [inline samples/OracleProvider/src/OracleProvider/Update/Internal/OracleModificationCommandBatch.cs]: Too many blank lines 😁

**Comment by AndriySvyryd** [comment]
@ralmsdeveloper The tests are still failing for me even with the newest changes. Do all of them pass for you?

**Comment by ralmsdeveloper** [comment]
Yes everything!
I was careful not to go back!

I'll check again. to release this pr

**Comment by ralmsdeveloper** [comment]
@AndriySvyryd I made a small adjustment, make sure it's okay!

File: samples/OracleProvider/src/OracleProvider/Update/Internal/OracleModificationCommandBatch.cs (hunk 12 above)
Comment (ralmsdeveloper) [reply samples/OracleProvider/src/OracleProvider/Update/Internal/OracleModificationCommandBatch.cs]: I did not realize that, but I corrected and sent!

**Comment by AndriySvyryd** [comment]
@ralmsdeveloper For example when I run `OracleEndToEndTest.Can_save_changes_in_tracked_entities()` I get this exception with your changes:
> ---- System.ArgumentException : Invalid parameter binding
Parameter name: cur
	C:\enlistments\Universe\EntityFramework\src\EFCore.Relational\Update\ReaderModificationCommandBatch.cs(302,0): at Microsoft.EntityFrameworkCore.Update.ReaderModificationCommandBatch.<ExecuteAsync>d__32.MoveNext()

We are using Oracle Database 12c Release 2 (12.2.0.1.0) - Standard Edition 2, see https://github.com/aspnet/EntityFrameworkCore/tree/dev/samples/OracleProvider

**Comment by ralmsdeveloper** [comment]
Okay @AndriySvyryd , I'm seeing this now, to release this PR!

**Comment by ralmsdeveloper** [comment]
I'm going to close this PR, because I'm updating my repositories, when I'm finished, I'll send the PR with the fixes!

I already made all the corrections for the implementation!

@AndriySvyryd

**Comment by ralmsdeveloper** [comment]
Moved to #10071

**Pull request closed without merging**